### PR TITLE
Update TZInfo to 2, with backwards compatibility option

### PR DESCRIPTION
**Pull request by jaredbeck**
### Summary
    
In the discussion in https://github.com/rails/rails/pull/34799 it was decided that ActiveSupport should upgrade to TZInfo 2, but provide a configuration option preserving backward compatibility with the output of TZInfo 1. I've chosen to call that option `tzinfo_compatibility_version`, but of course I'm open to other suggestions.

### CODE REVIEW

File: activesupport/lib/active_support/values/time_zone.rb [hunk 1]
```diff
@@ -507,10 +516,21 @@ def yesterday
...
+    # -- if you want an ActiveSupport::TimeWithZone instance, use
+    # Time#in_time_zone() instead.
+    #
+    # As of tzinfo 2, utc_to_local returns a Time with a non-zero utc_offset.
     def utc_to_local(time)
-      tzinfo.utc_to_local(time)
+      t = tzinfo.utc_to_local(time)
+      case ActiveSupport.tzinfo_compatibility_version
+      when "1"
+        Time.utc(t.year, t.month, t.day, t.hour, t.min, t.sec, t.sec_fraction)
```
Comment (jaredbeck) [inline activesupport/lib/active_support/values/time_zone.rb]: Is there a better way to construct this `Time`? I tried a lot of random other stuff 😅

**Comment by kaspth** [comment]
Finally got around to looking at this, and wanted to explore a slightly simpler config. Did so in https://github.com/rails/rails/commit/8b895b0647462374a6655e75acc64b6b30feaa7e

But there's loads of test failures, which I just don't understand.

For instance, test/core_ext/time_with_zone_test.rb:1306 fails when trying to locate the zone with `-9.hours` through `ActiveSupport::TimeZone[]`. That method finds a zone via its `utc_offset`: https://github.com/rails/rails/blob/fba1064153d8e2f4654df7762a7d3664b93e9fc8/activesupport/lib/active_support/values/time_zone.rb#L240 

But comparing the different configs with v1 it seems like even the v1 compatibility version doesn't match:

```
true, v2
[-43200, -39600, -39600, -36000, -32400, -25200, -25200, -25200, -28800, -28800, -21600, -21600, -21600, -21600, -25200, -21600, -18000, -21600, -18000, -18000, -14400, -18000, -14400, -18000, -14400, -14400, -14400, -10800, -10800, -10800, -10800, -14400, -12600, -7200, -3600, -3600, 0, 0, 0, 0, 0, 0, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 0, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 10800, 10800, 10800, 10800, 10800, 10800, 10800, 10800, 12600, 14400, 14400, 14400, 14400, 14400, 14400, 14400, 16200, 18000, 18000, 18000, 18000, 19800, 19800, 19800, 19800, 19800, 20700, 21600, 21600, 21600, 21600, 23400, 25200, 25200, 25200, 25200, 25200, 28800, 28800, 28800, 28800, 28800, 28800, 28800, 28800, 28800, 32400, 32400, 32400, 32400, 32400, 34200, 36000, 36000, 36000, 36000, 34200, 36000, 36000, 39600, 36000, 39600, 39600, 39600, 36000, 43200, 43200, 43200, 43200, 46800, 46800, 43200, 45900, 46800]

false, v1
[-43200, -39600, -39600, -36000, -28800, -25200, -25200, -25200, -25200, -25200, -21600, -21600, -21600, -21600, -21600, -21600, -18000, -18000, -18000, -18000, -14400, -14400, -14400, -14400, -14400, -14400, -10800, -10800, -10800, -10800, -10800, -10800, -9000, -7200, -3600, -3600, 0, 0, 0, 0, 0, 0, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 10800, 10800, 10800, 10800, 10800, 10800, 10800, 10800, 12600, 14400, 14400, 14400, 14400, 14400, 14400, 14400, 16200, 18000, 18000, 18000, 18000, 19800, 19800, 19800, 19800, 19800, 20700, 21600, 21600, 21600, 21600, 23400, 25200, 25200, 25200, 25200, 25200, 28800, 28800, 28800, 28800, 28800, 28800, 28800, 28800, 28800, 32400, 32400, 32400, 32400, 32400, 34200, 36000, 36000, 36000, 36000, 37800, 39600, 39600, 39600, 39600, 39600, 39600, 39600, 39600, 43200, 43200, 43200, 46800, 46800, 46800, 46800, 49500, 50400]

actual v1
[-43200, -39600, -39600, -36000, -32400, -28800, -28800, -25200, -25200, -25200, -25200, -21600, -21600, -21600, -21600, -21600, -21600, -18000, -18000, -18000, -18000, -18000, -14400, -14400, -14400, -14400, -14400, -14400, -12600, -10800, -10800, -10800, -10800, -7200, -3600, -3600, 0, 0, 0, 0, 0, 0, 0, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 3600, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 7200, 10800, 10800, 10800, 10800, 10800, 10800, 10800, 10800, 12600, 14400, 14400, 14400, 14400, 14400, 14400, 14400, 16200, 18000, 18000, 18000, 18000, 19800, 19800, 19800, 19800, 19800, 20700, 21600, 21600, 21600, 21600, 23400, 25200, 25200, 25200, 25200, 25200, 28800, 28800, 28800, 28800, 28800, 28800, 28800, 28800, 28800, 32400, 32400, 32400, 32400, 32400, 34200, 34200, 36000, 36000, 36000, 36000, 36000, 36000, 36000, 36000, 39600, 39600, 39600, 39600, 43200, 43200, 43200, 43200, 43200, 45900, 46800, 46800, 46800]
```

If I enable the v2 version many of the test failures pass by the way.

I don't get it. But maybe I'm completely off here. Any ideas? I'm surprised the build passed way back when you wrote your version and no tests hit any issues.

Any ideas? Happy to get this in and don't want it to wait such a long time again. @jonathanhefner curious if you'd happen to have any ideas here?

**Comment by jaredbeck** [comment]
> Happy to get this in and don't want it to wait such a long time again.

Great, yeah, I'd completely forgotten this PR existed.

> I'm surprised the build passed way back when you wrote your version and no tests hit any issues.

f5d00991f5861c348732764c2d0d1dba32c3c1ca merges `master`, causing a build, so we can see if it's still passing.

> .. wanted to explore a slightly simpler config. Did so in 8b895b0 ..

Comparing `use_tzinfo2_format` to `tzinfo_compatibility_version`, the later has the (minor) advantage of supporting further versions (eg. 3) in the future. There are other pros and cons, I'm sure, that's just the first thing that came to mind.

Thanks for the review.

**Comment by jonathanhefner** [comment]
> @jonathanhefner curious if you'd happen to have any ideas here?

@kaspth Not sure if you meant to tag me, but I dug into this a bit.  The reason the tests were passing previously but are failing now is because Daylight Savings Time started on March 8th.  Before DST the values of `base_utc_offset` and `observed_utc_offset` were the same.  Now they differ by 3600 (as do the mismatched offsets in the arrays you posted).

The reason some tests unexpectedly pass when using the v2 flag is that the condition for choosing `base_utc_offset` vs `observed_utc_offset` is, I believe, backwards:

https://github.com/rails/rails/blob/8b895b0647462374a6655e75acc64b6b30feaa7e/activesupport/lib/active_support/values/time_zone.rb#L307-L308

The [previous code](https://github.com/rails/rails/blob/fba1064153d8e2f4654df7762a7d3664b93e9fc8/activesupport/lib/active_support/values/time_zone.rb#L308) uses `TZInfo::TimezonePeriod#utc_offset` which [ignores DST](https://github.com/tzinfo/tzinfo/blob/048bad7dab90f19d408a0655033ee05eaba9645b/lib/tzinfo/timezone_period.rb#L41-L55).  Therefore, `base_utc_offset` should be used when mimicking v1 behavior.

Additionally, however, I'm not sure if using `observed_utc_offset` is ever correct (for `AS::TimeZone#utc_offset`).  Doing so means that TimeZone offsets will vary depending the on date.  I believe that is the cause of the [particular failure you mentioned](https://buildkite.com/rails/rails/builds/67589#e3e9e751-b271-4e42-8855-7a4668d29984/1003-1114).  (-9.hours == -32400 seconds, which is not an option in the "false, v1" array you posted, which was populated using `observed_utc_offset`.)  Is there a reason to desire this behavior?

After removing the conditional and always using `base_utc_offset`, all tests are passing in #38736.

**Comment by kaspth** [comment]
> Comparing `use_tzinfo2_format` to `tzinfo_compatibility_version`, the later has the (minor) advantage of supporting further versions (eg. 3) in the future. There are other pros and cons, I'm sure, that's just the first thing that came to mind.

We don't want that though. I'd rather we get people on Rails 6.1 then deprecate the backwards option and let people move to TZInfo2 (unless there's some extreme incompatibilities that force us not to).

@jonathanhefner perfect! yeah, I was curious to see if you'd be interested in looking and you were, thanks 🙌 I merged your PR into the tzinfo2 branch, because it's more correct than what was there before. I think I'll need to look into things again, to see what else would be missing for backwards compatibility or are we covered enough with the lone `use_tzinfo2_format` use?

**Comment by jonathanhefner** [comment]
@kaspth Looking through the TZInfo 2.0 [changelog](https://github.com/tzinfo/tzinfo/blob/master/CHANGES.md#version-200---26-dec-2018), I think Rails itself is covered.  I grepped for most of the changed and removed items in the changelog, and we seem to be handling them.

<details>
  <summary>grep for changed and removed TZInfo items</summary>

  ```bash
  $ grep "transitions_up_to|offsets_up_to|TimezonePeriod|RubyDataSource|ZoneinfoDataSource|InvalidZoneinfoDirectory|ZoneinfoDirectoryNotFound|DEFAULT_SEARCH_PATH|DEFAULT_ALTERNATE_ISO3166_TAB_SEARCH_PATH|Country[.]data_source|Timezone[.]data_source|create_default_data_source|process_search_path|Timezone[.]get_proxies|TransitionDataTimezoneInfo|timezone_identifiers" **/*.rb

  activesupport/lib/active_support/time_with_zone.rb:    # Returns the underlying TZInfo::TimezonePeriod.
  activesupport/test/core_ext/time_with_zone_test.rb:    assert_kind_of TZInfo::TimezonePeriod, @twz.instance_variable_get("@period")
  activesupport/test/time_zone_test.rb:    assert_kind_of TZInfo::TimezonePeriod, zone.period_for_local(Time.utc(2000))


  $ grep "valid_for_utc|utc_after_start|utc_before_end|valid_for_local|local_after_start|local_before_end|utc_total_offset_rational|local_end|local_end_time|local_start|local_start_time|Timezone[.]us_zones|Timezone[.]us_zone_identifiers" **/*.rb

  # nothing
  ```
</details>

I think the main issue is if existing app code has chosen to interact with the TZInfo API directly.  Rails mostly abstracts away from TZInfo, but we do [provide access](https://github.com/rails/rails/blob/6a41940a965241ede9a77837169d6a329adf13d7/activesupport/lib/active_support/values/time_zone.rb#L291) to the underlying `TZInfo::Timezone` if the user wants it.  Another spot where we expose the TZInfo API is [`AS::TimeWithZone#period`](https://github.com/rails/rails/blob/6a41940a965241ede9a77837169d6a329adf13d7/activesupport/lib/active_support/time_with_zone.rb#L72-L75), which returns a `TZInfo::TimezonePeriod`.  `TZInfo::TimezonePeriod` had [several methods removed](https://github.com/tzinfo/tzinfo/blame/master/CHANGES.md#L138-L142), but perhaps those methods aren't widely used.

**Comment by kaspth** [comment]
Resolved in bf34c808f9, thanks both of you 🙏

**Comment by ghiculescu** [comment]
Should we add a note about this to https://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-6-0-to-rails-6-1 ?

**Pull request closed without merging**